apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/RealCache.java

```diff
@@ -1,77 +1,94 @@
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 
 import static com.apollographql.android.api.graphql.util.Utils.checkNotNull;
 
-public final class RealCache implements Cache {
+public final class RealCache implements Cache, ReadableCache, WriteableCache {
   private final CacheStore cacheStore;
   private final CacheKeyResolver cacheKeyResolver;
   private final ReadWriteLock lock;
-  private final Map<RecordChangeSubscriber, Set<String>> subscribers;
+  private final Set<RecordChangeSubscriber> subscribers;
 
   public RealCache(@Nonnull CacheStore cacheStore, @Nonnull CacheKeyResolver cacheKeyResolver) {
     this.cacheStore = cacheStore;
     this.cacheKeyResolver = cacheKeyResolver;
     this.lock = new ReentrantReadWriteLock();
-    this.subscribers = new WeakHashMap<>();
+    this.subscribers = Collections.newSetFromMap(new WeakHashMap<RecordChangeSubscriber, Boolean>());
   }
 
   @Override @Nonnull public ResponseNormalizer responseNormalizer() {
     return new ResponseNormalizer(cacheKeyResolver);
   }
 
-  @Override public void subscribe(RecordChangeSubscriber subscriber, Set<String> dependentKeys) {
-    subscribers.put(subscriber, dependentKeys);
+  @Override public synchronized void subscribe(RecordChangeSubscriber subscriber) {
+    subscribers.add(subscriber);
   }
 
-  @Override public void unsubscribe(RecordChangeSubscriber subscriber) {
+  @Override public synchronized void unsubscribe(RecordChangeSubscriber subscriber) {
     subscribers.remove(subscriber);
   }
 
-  @Override public void write(@Nonnull Collection<Record> recordSet) {
-    Set<String> changedKeys = Collections.emptySet();
-    lock.writeLock().lock();
-    try {
-      changedKeys = cacheStore.merge(checkNotNull(recordSet, "recordSet == null"));
-    } finally {
-      lock.writeLock().unlock();
+  @Override public void publish(@Nonnull Set<String> changedKeys) {
+    checkNotNull(changedKeys, "changedKeys == null");
+    if (changedKeys.isEmpty()) {
+      return;
     }
-    Map<RecordChangeSubscriber, Set<String>> iterableSubscribers = new LinkedHashMap<>(subscribers);
-    for (Map.Entry<RecordChangeSubscriber, Set<String>> subscriberEntry : iterableSubscribers.entrySet()) {
-      if (!Utils.areDisjoint(subscriberEntry.getValue(), changedKeys)) {
-        subscriberEntry.getKey().onDependentKeysChanged();
-      }
+    Set<RecordChangeSubscriber> iterableSubscribers;
+    synchronized (this) {
+      iterableSubscribers = new LinkedHashSet<>(subscribers);
+    }
+    for (RecordChangeSubscriber subscriber : iterableSubscribers) {
+      subscriber.onCacheKeysChanged(changedKeys);
     }
   }
 
-  @Override public Record read(@Nonnull String key) {
-    lock.readLock().lock();
-    try {
-      return cacheStore.loadRecord(checkNotNull(key, "key == null"));
-    } finally {
-      lock.readLock().unlock();
-    }
+  @Override public <R> Transaction<ReadableCache, R> readTransaction() {
```

Comment:
Not sure if it makes sense, but we can simplify API even more, so instead returning the transaction why don't pass Transactional as a parameter here:

```
@Override public <R> R executeWithReadTransaction(Transactional<ReadableCache, R> transactional) {
  try {
    lock.readLock().lock();
    return transactional.call(RealCache.this);
  } finally {
    lock.readLock().unlock();
  }
}
```

The same for write. In this case we can rid off one indirection `Transaction` class at all.